File: python/sorting/shell_sort.py
from .sort import SortType, Sort


class ShellSort(Sort):
    def sort(self, arr: list[int], type=SortType.asc):
        length: int = len(arr)

        gap: int = 1
        while gap < length / 9:
            gap = gap * 3 + 1

        while gap > 0:
            for i in range(gap, length):
                j: int = i
                while j >= gap and super().compare(arr[j - gap], arr[j], type):
                    arr[j - gap], arr[j] = arr[j], arr[j - gap]
                    j -= 1

            gap = int(gap / 3)

        return arr

File: typescript/sorting/shell_sort.ts
import Sort, { SortType, SORT_TYPE } from "./sort.ts";

class ShellSort extends Sort {
  sort(arr: number[], type: SortType = SORT_TYPE.asc): number[] {
    const length: number = arr.length;

    let gap: number = 1;
    while (gap < length / 9) {
      gap = gap * 3 + 1;
    }

    while (gap > 0) {
      for (let i = gap; i < length; i++) {
        let j = i;
        while (j >= gap && super.compare(arr[j - gap], arr[j], type)) {
          [arr[j - gap], arr[j]] = [arr[j], arr[j - gap]];
          j--;
        }
      }

      gap = gap / 3;
    }

    return arr;
  }
}

export default ShellSort;
